Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,183 @@
+;;; innermost child.
+;;;
+;;; Other pointer events result in finding a common ancestor of the
+;;; old port-pointer-sheet and the innermost sheet under the
+;;; pointer. First we synthetize exit events and then enter events.
+(defun synthetize-boundary-events (port event)
+  (labels ((synthetize-enter (sheet)
+             (dispatch-event-copy sheet event 'pointer-enter-event)
+             (setf (port-pointer-sheet port) sheet))
+           (synthetize-exit (sheet)
+             (setf (port-pointer-sheet port) nil)
+             (dispatch-event-copy sheet event 'pointer-exit-event))
+           (map-within-mirror (function sheet x y)
+             (funcall function sheet)
+             (map-over-sheets-containing-position
+              (lambda (child)
+                (unless (sheet-direct-mirror child)
+                  (multiple-value-bind (child-x child-y)
+                      (untransform-position (sheet-transformation child) x y)
+                    (map-within-mirror function child child-x child-y))))
+              sheet x y)))
+    ;; FIXME clx generates uneven number of ungrab-enter/ungrab-leave
+    ;; events what may disrupt the enter/exit event balance for a
+    ;; sheet. Not distributing such events makes overall UX worse, for
+    ;; instance user needs first to click on a window to press an
+    ;; armed push-button. What's the right thing? -- jd 2019-08-28
+    (typecase event
+      ;; (pointer-ungrab-enter-event (warn "ungrab enter event"))
+      ;; (pointer-ungrab-leave-event (warn "ungrab leave event"))
+      (pointer-enter-event
+       (let ((sheet (event-sheet event)))
+         (get-pointer-position (sheet event)
+           (map-within-mirror #'synthetize-enter sheet x y))))
+      (pointer-exit-event
+       (when-let ((sheet (event-sheet event))
+                  (child (port-pointer-sheet port)))
+         (when (eql sheet (sheet-mirrored-ancestor child))
+           (loop
+              for over-sheet = child then (sheet-parent over-sheet)
+              do (synthetize-exit over-sheet)
+              until (eql over-sheet sheet)))))
+      (otherwise
+       (flet ((common-ancestor (sheet-a sheet-b)
+                (loop
+                   (cond ((or (null sheet-a) (null sheet-b)
+                              (sheet-direct-mirror sheet-a)
+                              (sheet-direct-mirror sheet-b))
+                          (return-from common-ancestor nil))
+                         ((sheet-ancestor-p sheet-b sheet-a)
+                          (return-from common-ancestor sheet-a))
+                         (t (setf sheet-a (sheet-parent sheet-a)))))))
+         (let* ((old-sheet (port-pointer-sheet port))
+                (new-sheet (compute-pointer-event-sheet event))
+                (common-ancestor (common-ancestor old-sheet new-sheet)))
+           ;; distribute exit events (innermost first)
+           (do ((sheet old-sheet (sheet-parent sheet)))
+               ((or (null sheet)
```

Comment:
Is `(null sheet)` needed? If `common-ancestor` is non-`nil`, it is encountered before `sheet` becomes `nil`. If `common-ancestor` is `nil`, `(null sheet)` is the same as `(eq sheet common-ancestor)`.

Also, it is mildly confusing that basically the same operation is written using `loop` in line 266 and using `do` here.

Reply:
I think it was needed, but I'll think about it later. do is used to reflect the similarity between enter/leave synthesize (but in reverse order) (not that in the first we distribute thing eagerly and in the other we first push and then we use the order on the list).